pkg/policy/mapstate.go

```diff
@@ -100,3 +104,79 @@ func (keys MapState) AllowAllIdentities(ingress, egress bool) {
 		keys[keyToAdd] = MapStateEntry{}
 	}
 }
+
+// MapChanges collects updates to the endpoint policy on the
+// granularity of individual mapstate key-value pairs for both adds
+// and deletes. 'mutex' must be held for any access.
+type MapChanges struct {
+	mutex   lock.Mutex
+	adds    MapState
+	deletes MapState
+}
+
+// AccumulateMapChanges accumulates the given changes to the
+// MapChanges, updating both maps for each add and delete, as
+// applicable.
+func (mc *MapChanges) AccumulateMapChanges(adds, deletes []identity.NumericIdentity,
+	port uint16, proto uint8, direction trafficdirection.TrafficDirection) {
+	key := Key{
+		// The actual identity is set in the loops below
+		Identity: 0,
+		// NOTE: Port is in host byte-order!
+		DestPort:         port,
+		Nexthdr:          proto,
+		TrafficDirection: direction.Uint8(),
+	}
+	value := MapStateEntry{
+		ProxyPort: 0, // Will be updated by the caller when applicable
+	}
+
+	log.WithFields(logrus.Fields{
+		logfields.AddedPolicyID:    adds,
+		logfields.DeletedPolicyID:  deletes,
+		logfields.Port:             port,
+		logfields.Protocol:         proto,
+		logfields.TrafficDirection: direction,
+	}).Debug("AccumulateMapChanges")
+
+	mc.mutex.Lock()
+	if len(adds) > 0 {
+		if mc.adds == nil {
+			mc.adds = make(MapState)
+		}
+		for _, id := range adds {
+			key.Identity = id.Uint32()
+			mc.adds[key] = value
+			// Remove a potential previously deleted key
+			if mc.deletes != nil {
+				delete(mc.deletes, key)
+			}
+		}
+	}
+	if len(deletes) > 0 {
+		if mc.deletes == nil {
+			mc.deletes = make(MapState)
+		}
+		for _, id := range deletes {
+			key.Identity = id.Uint32()
+			mc.deletes[key] = value
+			// Remove a potential previously added key
+			if mc.adds != nil {
+				delete(mc.adds, key)
+			}
+		}
+	}
+	mc.mutex.Unlock()
+}
+
+// GetMapChanges transfers the changes from MapChanges to the caller.
+// May return nil maps.
+func (mc *MapChanges) GetMapChanges() (adds, deletes MapState) {
```

Comment:
```suggestion
func (mc *MapChanges) StealMapChanges() (adds, deletes MapState) {
```

or (wait for it)
```suggestion
func (mc *MapChanges) ConsumeMapChanges() (adds, deletes MapState) {
```
? 

This would be a bit more indicative that it's a write operation. That said, feel free to disagree.

Reply:
You guys know what naming convention I am for here ;)

Reply:
I went with `ConsumeMapChanges()`, thanks for the suggestion :-)